backend/drm.c

```diff
@@ -0,0 +1,337 @@
+/*
+ * Twin - A Tiny Window System
+ * Copyright (c) 2024 National Cheng Kung University, Taiwan
+ * All rights reserved.
+ */
+
+#include <fcntl.h>
+#include <libdrm/drm.h>
+#include <libdrm/drm_mode.h>
+#include <linux/kd.h>
+#include <linux/vt.h>
+#include <stdlib.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <twin.h>
+#include <xf86drm.h>
+#include <xf86drmMode.h>
+
+#include "linux_input.h"
+#include "linux_vt.h"
+#include "twin_backend.h"
+
+#define DRM_DRI_NAME "DRI_CARD"
+#define DRM_DRI_DEFAULT "/dev/dri/card0"
+#define SCREEN(x) ((twin_context_t *) x)->screen
+#define RES(x) ((twin_drm_t *) x)->res
+#define CONN(x) ((twin_drm_t *) x)->conn
+#define CRTC(x) ((twin_drm_t *) x)->crtc
+#define PRIV(x) ((twin_drm_t *) ((twin_context_t *) x)->priv)
+
+typedef struct {
+    twin_screen_t *screen;
+
+    /* Linux input system */
+    void *input;
+
+    /* Linux virtual terminal (VT) */
+    int vt_fd;
+    int vt_num;
+    bool vt_active;
+
+    /* DRM driver */
+    int width, height;
+    int drm_dri_fd;
+    drmModeResPtr res;
+    drmModeConnectorPtr conn;
+    drmModeCrtcPtr crtc;
+    uint32_t fb_id;
+    uint32_t *fb_base;
+    size_t fb_len;
+    // int display_width;
+} twin_drm_t;
+
+static void _twin_drm_put_span(twin_coord_t left,
+                               twin_coord_t top,
+                               twin_coord_t right,
+                               twin_argb32_t *pixels,
+                               void *closure)
+{
+    twin_drm_t *tx = PRIV(closure);
+
+    if (tx->fb_base == MAP_FAILED)
+        return;
+
+    twin_coord_t width = right - left;
+    int line_length = 4 * tx->width;
+    off_t off = line_length * top + 4 * left;
+    unsigned char *dest = (unsigned char *) tx->fb_base + off;
```

Comment:
Replace `unsigned char` with `uint8_t` for consistency and use `uintptr_t` for pointer arithmetic. That is:
```C
uint8_t *dest = (uint8_t *)((uintptr_t) tx->fb_base + off);
```